File: #0077.combinations.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {

        vector<vector<int>> result;
        vector<int> backtrack(k, 0);
        int index = 0;
        while (index >= 0) {
            ++backtrack[index];
            if (backtrack[index] > n) {
                --index;
            } else if (index == k - 1) {
                result.push_back(backtrack);
            } else {
                ++index;
                backtrack[index] = backtrack[index - 1];
            }
        }
        return result;
    }
};

File: #0077.combinations.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function(n, k) {
    let result = [];

    function backtrack(curr = [], index = 1) {
        if (curr.length === k) {
            return result.push(curr);
        }

        for (let i = index; i < n + 1; i++) {
            backtrack([...curr, i], i + 1);
        }
    }

    backtrack();
    return result;
};
